### PR TITLE
feat(template): allow variable references in include/exclude config

### LINKED ISSUE #4619: 0.13: [Bug]: action-variables can't be used in action's excludes or includes
## Garden Bonsai (0.13) Bug
I can't use action-variables in excludes or includes.

### Current Behavior

```
% garden validate
Error processing config for Deploy action test2:
Invalid template string (${var.anyVar}): Could not find key anyVar under var. Available keys: (none).
```

### Expected behavior

Successful validation

### Reproducible example

```
kind: Project
apiVersion: garden.io/v1
name: repro
environments:
 - name: default
providers:
 - name: local-kubernetes

---
kind: Deploy
type: exec
name: test2
build: test
exclude: "${var.anyVar}"
variables:
  anyVar:
    - "*/**"
```

### Workaround

1. Use project-level variables if possible
2. Do not use variables (e.g repeat code)

### Suggested solution(s)

Allow using action-based variables. We could allow that by lazily evaluating variables and only failing if they refer to a context that's not available yet in excludes/ includes

### Your environment

* **OS:** macOS
* **How I'm running Kubernetes:** EKS

`garden version` 0.13.2

### PR DESCRIPTION
**Pull request by vvagaytsev**
**What this PR does / why we need it**:
See individual commits. There is some refactored code. The feature itself is shipped as a separate commit.

**Which issue(s) this PR fixes**:

Fixes #4619

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by vvagaytsev** [comment]
Huge thanks to @thsig for pairing on this! 👍

File: core/src/config/template-contexts/actions.ts [hunk 1]
```diff
@@ -65,10 +72,10 @@ export class ActionConfigContext extends TemplatableConfigContext {
   @schema(ActionConfigThisContext.getSchema().description("Information about the action currently being resolved."))
   public this: ActionConfigThisContext
 
-  constructor(garden: Garden, config: ActionConfig | WorkflowConfig, params: ActionConfigThisContextParams) {
+  constructor({ garden, config, thisContextParams, variables }: ActionConfigContextParams) {
     super(garden, config)
-    const { name, mode } = params
-    this.this = new ActionConfigThisContext(this, params)
+    this.this = new ActionConfigThisContext(this, thisContextParams)
```
Comment (thsig) [inline core/src/config/template-contexts/actions.ts:77]: The thisness of the this itself, as manifested through its own instrumentality.

Very Heidegger. Haha.

**Review by thsig — APPROVED** [review]

File: core/src/config/template-contexts/actions.ts (hunk 1 above)
Comment (vvagaytsev) [reply core/src/config/template-contexts/actions.ts:77]: Yeah, that's the necessary evil here :)

**Comment by shumailxyz** [comment]
Cool stuff 💯

**Pull request merged**